include/zim/archive.h

```diff
@@ -544,7 +544,7 @@ namespace zim
   class LIBZIM_API Archive::iterator
   {
     public:
-      using iterator_category = std::bidirectional_iterator_tag;
+      using iterator_category = std::input_iterator_tag;
```

Comment:
I disagree with this change for two reasons:

1. `Archive::iterator` is conceptually a bidirectional iterator (although it doesn't comply with all the requirements that STL has chosen to impose on its category of bidirectional iterators).
2. `Archive::iterator` is significantly more capable than STL's *`LegacyInputIterator`* since it can be used in multi-pass algorithms. *`LegacyInputIterator`* is about *other iterators* being invalidated, whereas the issue with `Archive::iterator` is that the `reference` obtained via `operator*` is invalidated as soon as the iterator is moved (unless `Archive::iterator::reference` is defined to be a value rather than a c++ reference).

The same argument goes for similar downshifts in `SearchIterator` and `SuggestionIterator`

Reply:
From [bidirectionalIterator](https://en.cppreference.com/w/cpp/named_req/BidirectionalIterator), a Legacy Bidirectional iterator is also a *[LegacyForwardIterator](https://en.cppreference.com/w/cpp/named_req/ForwardIterator)*

But from the note in ForwadIterator documentation, a *LegacyForwardIterator* reference must return a reference:

> Unlike the [std::forward_iterator](https://en.cppreference.com/w/cpp/iterator/forward_iterator) concept, the LegacyForwardIterator requirements requires dereference to return a reference.


(And `std::forward_iterator` is c++20 only)

I think we are stuck here. STL impose us some constraints we cannot provide as we create the object on demand instead of using one object stored in the container.
(I would be glad you prove me I'm wrong)

Declaring the iterator as STL LegacyInputIterator will limits only the STL. We may still document that iterator is kind of Bidirectional iterator but that referenced object must not be used after modification of iterator itself.

Reply:
@veloman-yunkan Are we agree on this ? At least as a "non blocker issue, still to fix" ?
This PR is waited as the current global CI is broken and we need this PR to fix it.

Reply:
> (I would be glad you prove me I'm wrong)

Unfortunately, I have to disappoint you - you are perfectly right

> @veloman-yunkan Are we agree on this ?

With all the *input* provided let's move *forward* with the current code but explain in a comment (summarizing this discussion) the reason for `std::input_iterator_tag` in `Archive::iterator`, `SearchIterator` and `SuggestionIterator`.

Reply:
Done in a fixup commit.